### PR TITLE
Eliminate clang warnings in document

### PR DESCRIPTION
**Pull request by toregge**
@vekterli: please review

### CODE REVIEW

File: document/src/vespa/document/base/idstring.cpp [hunk 1]
```diff
@@ -153,8 +153,13 @@ inline const char *
 fmemchr(const char * s, const char * e)
 {
     while (s+15 < e) {
+#ifdef __clang__
+        v16qi tmpCurrent = __builtin_ia32_lddqu(s);
+        v16qi tmp0       = tmpCurrent == _G_zero;
+#else
         v16qi tmpCurrent = __builtin_ia32_loaddqu(s);
         v16qi tmp0       = __builtin_ia32_pcmpeqb128(tmpCurrent, _G_zero);
+#endif
```
Comment (vekterli) [inline document/src/vespa/document/base/idstring.cpp]: We should rewrite this later to use more standardized intrinsics such as `_mm_loadu_si128`, `_mm_movemask_epi8` etc. I did that locally at some point when running the doc id parser code through libFuzzer, will try to dig it up.

File: document/src/vespa/document/fieldvalue/structuredfieldvalue.cpp [hunk 2]
```diff
@@ -18,15 +18,13 @@ using namespace fieldvalue;
...
-      _iterator(),
+    : _iterator(),
       _field(0)
 {
 }
 
-StructuredFieldValue::Iterator::Iterator(const StructuredFieldValue& owner, const Field* first)
-    : _owner(&const_cast<StructuredFieldValue&>(owner)),
-      _iterator(owner.getIterator(first).release()),
+StructuredFieldValue::Iterator::Iterator([[maybe_unused]] const StructuredFieldValue& owner, const Field* first)
```
Comment (vekterli) [inline document/src/vespa/document/fieldvalue/structuredfieldvalue.cpp]: `owner` is used as part of `_iterator` construction below, remove `[[maybe_unused]]`?

File: document/src/vespa/document/base/idstring.cpp [hunk 3]
```diff
@@ -153,8 +153,13 @@ inline const char *
 fmemchr(const char * s, const char * e)
 {
     while (s+15 < e) {
+#ifdef __clang__
+        v16qi tmpCurrent = __builtin_ia32_lddqu(s);
```
Comment (baldersheim) [inline document/src/vespa/document/base/idstring.cpp]: Finnes det ikke noe som fungerer for begge ?

**Review by baldersheim — APPROVED** [review]

**Review by vekterli — APPROVED** [review]
:+1: see comment on parameter attribute

File: document/src/vespa/document/base/idstring.cpp (hunk 3 above)
Comment (vekterli) [reply document/src/vespa/document/base/idstring.cpp]: Something like this _should_ be portable (warning: I have not verified 100% identical semantics):
```c++
#include <emmintrin.h>
(...)
inline const char *
fmemchr(const char * s, const char * e)
{
    const auto sep = _mm_set1_epi8(':');
    while (s+15 < e) {
        auto current = _mm_loadu_si128(reinterpret_cast<const __m128i*>(s));
        auto mask = _mm_movemask_epi8(_mm_cmpeq_epi8(current, sep));
        if (__builtin_expect(mask, 1)) {
            return s + lsbIdx(uint32_t(mask));
        }
        s+=16;
    }
(...)
```

**Review by vekterli — APPROVED** [review]

File: document/src/vespa/document/base/idstring.cpp (hunk 1 above)
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.cpp]: Do not forget cleanup.

File: document/src/vespa/document/base/idstring.cpp (hunk 3 above)
Comment (vekterli) [reply document/src/vespa/document/base/idstring.cpp]: Both GCC and clang seems to happily eat the above for dinner, at least: https://godbolt.org/z/08UbwY

**Pull request merged**